### DISCUSSION TITLE
Windows VM scripts

### DISCUSSION THREAD

**Discussion by Schinken6** [community-scripts/ProxmoxVE#765] [Q&A]
Hello,
I am quite new to this whole self hosting thing and was just searching for a windows VM script but couldn't find one. Is there a reason there is no script for it like for not being open source or hate against Windows in general?

**Comment by alexanderjerome** [comment]
The lack of Windows scripts is mostly due to licensing issues. Microsoft doesn't provide permanent download URLs for Windows ISOs, and automating Windows installations can be legally tricky.

That said, it's totally possible to make a script work with user-provided ISOs. You could have a script that sets up the VM parameters (q35, UEFI, TPM), handles VirtIO drivers, and even helps with answer files for unattended installation. The user would just need to bring their own Windows media.

Just my thoughts on why it's not in there yet - might be worth exploring if people are interested in it.

**Reply by Schinken6** [reply]
Thanks for the reply. That could totally work.
I noticed that too. When you download the Windows 11 ISO. Windows states that the link for downloading is only valid for 24 hours. So you would have to make a script somehow to automate the process of recreating the link or as you said just let the user predownload it and point the automation script to the ISO.
I noticed when I downloaded Windows Server 2022 my link from a year ago still worked as I have written it into my documentation.
I am not sure about rules but I can post the link if you are curious.

**Reply by solidgonk** [reply]
I started developing something recently and figured I might check in here to see if there was some interest. I think that a core set of Windows scripts would benefit the Proxmox community, and could also expand use of the platform within the industry. 

In response to the concerns around distribution and stale download links, there is no reason why the script couldn't just as you to select an existing ISO from one of the chosen storage locations. Microsoft literally make their ISOs available to the public, and there are many tools that already pull them their cloud repositories.

You can see a static download link in this installer script for Ludus (which relies on Proxmox, but uses Ansible): https://github.com/badsectorlabs/ludus/blob/53a65d3ce8226b5ca3dea14dccae09e25d1fb55c/templates/win11-24h2-x64-enterprise-tpm/win11-24h2-x64-enterprise-tpm.pkr.hcl

The VM does require an Autounattend.xml file to be added to an ISO, which is easily done with a one-liner in mkisofs after the user has entered whatever customer configuration options they are prompted for. The Autounattend.xml file would also make the relevant  drivers available from the virtio ISO.

You can see an example here, also from the Ludus project: https://github.com/badsectorlabs/ludus/blob/53a65d3ce8226b5ca3dea14dccae09e25d1fb55c/templates/win11-24h2-x64-enterprise-tpm/Autounattend.xml

If there's interest, I'll keep working on it.
